### PR TITLE
PNG support overhaul

### PR DESCRIPTION
**Pull request by JulianGmp**
I wanted to add some PNG images but noticed some weren't working, specifically ones with with 8-Bit grayscale info and 8-Bit indexing (color type 0 and 3). So I decided to implement it.

One part I am a bit unsure about is how I handled the `colour_space` string. It's manually allocated and freed since it might be just 12 or around 2000 bytes long, and I thought that I shouldn't just make the 2k buffer in the cases where I don't even need it.
At times like this I miss `std::string`, but if you have a suggestion on how to do this better please let me know.

_Disclaimer: I'm sending this PR at 1 am this might as well just be garbage that seems to work._

## Changes
- Use `uint32_t` as opposed to `unsigned` wherever possible (368b79c)
- The PNG color types are now stored in an enum, as opposed to macros
  - the color types were also expanded with all types listed, for completeness
- the color type is now stored in the `png_info` struct, as opposed to `ncolours`.
- added `rgb_value` struct for storing the indexed colour palette data
- parse the PNG palette chunk for indexed PNG images
- output the indexed colour palette into the PDF stream
- add a check to see if the header chunk is not the first chunk to be read
- add `pdf_add_grayscale8` to allow adding raw grayscale images, for completeness*

*this also allows fairly easily adding 8-Bit BMP data by modifying `pdf_add_bmp_data`, but that's out of scope for this PR.

## Resources
These might be useful to understand what some of the code is even trying to do.
- The [PDF Reference](https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/pdf_reference_archive/PDFReference.pdf)
specifically chapter 8.6.6.3
- The [PNG Specification](https://www.w3.org/TR/2003/REC-PNG-20031110/)
specifically chapters 5 and 6.1.

### CODE REVIEW

**Comment by AndreRenaud** [comment]
Hi Julian,
Thanks for this - it looks really good.

I'd sort of like to split it into two PRs just for review purposes - one just the type changes (going to uin32_t etc...), and the other for the PNG stuff. I think the type change stuff is pretty much a no-brainer, and makes things better. But equally, you've done the work, so I'm not too fussed about this. If you'd rather stick with it all in this PR, that's fine as well.

The PNG thing I'm not quite clear on how it works with greyscale and palette'd images. Are you able to stick some example files into the `data` directory, and make `tests/main.c` output them? Superficially, it all looks good though.

Andre

File: pdfgen.c [hunk 1]
```diff
@@ -2561,59 +2643,115 @@ static int pdf_add_png_data(struct pdf_doc *pdf, struct pdf_object *page,
...
+                                "not divisbly by 3!");
+                    goto info_free;
+                }
+                palette_buffer_length = (size_t)(chunk_length / 3);
+                palette_buffer = (struct rgb_value *)malloc(
+                    palette_buffer_length * sizeof(struct rgb_value));
+                if (!palette_buffer) {
+                    pdf_set_err(pdf, -ENOMEM,
+                                "Could not allocate memory for indexed color "
+                                "palette (%d bytes)",
```
Comment (AndreRenaud) [inline pdfgen.c:2691]: `palette_buffer_length` is a `size_t`, so these should be `%zu` not `%d`

File: pdfgen.c [hunk 2]
```diff
@@ -2622,20 +2760,64 @@ static int pdf_add_png_data(struct pdf_doc *pdf, struct pdf_object *page,
...
+    case PNG_COLOR_INDEXED:
+        // Write the color palette to the color_palette buffer
+        // the 2KiB buffer length should never be exceeded, even if all 256
+        // palette indices have an assigned RGB value
+        const size_t buffer_size = 2048;
+        colour_space = calloc(buffer_size, sizeof(char));
+        int buffer_pos = snprintf(colour_space, buffer_size,
+                                  "[ /Indexed\r\n"
+                                  "  /DeviceRGB\r\n"
+                                  "  %d\r\n"
```
Comment (AndreRenaud) [inline pdfgen.c:2782]: `%zu` as above here too

File: pdfgen.c [hunk 3]
```diff
@@ -2622,20 +2760,64 @@ static int pdf_add_png_data(struct pdf_doc *pdf, struct pdf_object *page,
         goto info_free;
     }
 
-    /* RGB colored image */
+    char *colour_space = NULL;
```
Comment (AndreRenaud) [inline pdfgen.c:2763]: As you mentioned, string handling in C sort of sucks. I've got the `dstr` stuff in here in an attempt to mitigate it. So you can use:
```c
struct dstr colour_space = INIT_DSTR;
dstr_append(&colour_space, "/DeviceGray");
...
dstr_free(&colour_space)
```
If the string is < 64 bytes long, no dynamic allocation will actually occur (it'll just live on the stack). If it ends up being longer, it'll extend as needed

Comment (JulianGmp) [reply pdfgen.c]: thank you for the info, I've implemented it like this now.

**Comment by JulianGmp** [comment]
>The PNG thing I'm not quite clear on how it works with greyscale and palette'd images. Are you able to stick some example files into the data directory, and make tests/main.c output them? Superficially, it all looks good though.

I created some example files, you can check their colour mode with the `file` command or with `magick identify`..

As for storing them in the PDF:
The greyscale images are fairly simple, we only tell the PDF that we have one colour channel and pass `/DeviceGray`. The image data gets streamed in the same way as with RGB data.

The indexed data requires some more attention. We stream in the image data the same way as with the grey images (one 8-bit channel). The values in the image data refer to colour in the palette. In the PNG stream the palette is stored with the `PLTE` chunk, each color is a 24-bit RGB value, and its index is derived from its position within the stream.
Since we have one byte of index range we can have up 256 colors in the palette, but we may also have less than that (the example file has around 170 iirc).

When streaming the palette into PDF we define the color space as indexed.
The indexed value then maps to an actual color space (`DeviceRGB`, although PDF would technically allow us to also use CMYK for example).
The `174` defines the max index (i.e. the number of colors in the palette - 1).
After that we write all 175 RGB values in the `< >` as hex values (trimmed for readability here)
```
/Name /Image36
/Subtype /Image
/ColorSpace [ /Indexed
  /DeviceRGB
  174
  <FFAA33 88DD33 33FF33 AAFFAA FFFFFF 3355FF AACC33 BBBBFF...... >
]
\Width 128
and so on
```
The position of the RGB value within the `< >` also determine its index value, like it does in the PNG PLTE chunk. So we only need to write them in the same order as we read them.

It's important to note that the max index and the number of values need to match. I mistook the value for being the length as opposed to the max index (length-1) and my PDF viewer would refuse to show the image.

**Comment by JulianGmp** [comment]
I rebased from master and addressed your suggestions.

File: pdfgen.c [hunk 4]
```diff
@@ -2511,26 +2567,45 @@ int pdf_add_rgb24(struct pdf_doc *pdf, struct pdf_object *page, float x,
...
+                       float y, float display_width, float display_height,
+                       const uint8_t *data, uint32_t width, uint32_t height)
+{
+    struct pdf_object *obj;
+
+    obj = pdf_add_raw_grayscale8(pdf, data, width, height);
+    if (!obj)
+        return pdf->errval;
+
+    return pdf_add_image(pdf, page, obj, x, y, display_width, display_height);
```
Comment (AndreRenaud) [inline pdfgen.c:2580]: Does this now need a call to `get_img_display_dimensions`?

Comment (JulianGmp) [reply pdfgen.c:2580]: yeah, this PR branch needs to be rebased from master.

I'd have done this already but my laptop's cpu cooler has decided to die on me so I think I'll get back to this in a couple of days.

Comment (AndreRenaud) [reply pdfgen.c:2580]: Let me know if you'd like me to take this over and clean up the last small bits for merge. I think it's almost sorted.

Comment (JulianGmp) [reply pdfgen.c:2580]: Update: yeah it will probably take some time until I can get back to this, if you could finish this up that would be great.

Comment (AndreRenaud) [reply pdfgen.c:2648]: No problem. I also realised that your new greyscale support lets us trivially add in support for .pgm binary image files as well, so I've done that. That also means that the fuzz tester will be able to reach the new greyscale functions, which is also nice.

Comment (JulianGmp) [reply pdfgen.c:2648]: Yet another format, that's nice to see.
Thank you for finishing this up.

**Pull request merged**